INTMDB-587: New Feature: Backup Compliance Policy Support (#1127)
mongodb/terraform-provider-mongodbatlas: website/docs/d/backup_compliance_policy.html.markdown

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: backup_compliance_policy"
sidebar_current: "docs-mongodbatlas-datasource-backup-compliance-policy"
description: |-
    Provides a Backup Compliance Policy Datasource.
---

> # Data Source: mongodbatlas_backup_compliance_policy

Reviewer: After we sort out the resource docs page issues and functionality over all this will need to be updated.  Not going into details here as it will be redundant to resource.

> `mongodbatlas_backup_compliance_policy` provides a Backup Compliance Policy data source. An Atlas Backup Compliance Policy provides the current protection policy settings for the project. 

Reviewer 2: ```suggestion
`mongodbatlas_backup_compliance_policy` provides a Backup Compliance Policy data source. An Atlas Backup Compliance Policy provides the current protection policy settings for the project. Prevent any user, regardless of role, from modifying or deleting specific cluster configurations and backups. When enabled, the Backup Compliance Policy will be applied as the minimum policy for all clusters and can only be disabled by contacting MongoDB support. Only supported for clusters M10 or higher.
```

* `frequency_type` - Frequency associated with the backup policy item. For monthly policies, the frequency type is defined as `monthly`. Note that this is a read-only value and not required in plan files - its value is implied from the policy resource type.
* `frequency_interval` - Desired frequency of the new backup policy item specified by `frequency_type` (monthly in this case). The supported values for weekly policies are 
  * `1` through `28` where the number represents the day of the month i.e. `1` is the first of the month and `5` is the fifth day of the month.
  * `40` represents the last day of the month (depending on the month).
* `retention_unit` - Scope of the backup policy item: `days`, `weeks`, or `months`.
* `retention_value` - Value to associate with `retention_unit`. Monthly policy must have retention days of at least 31 days or 5 weeks or 1 month. Note that for less frequent policy items, Atlas requires that you specify a retention period greater than or equal to the retention period specified for more frequent policy items. For example: If the weekly policy item specifies a retention of two weeks, the montly retention policy must specify two weeks or greater.

> For more information see: [MongoDB Atlas API Reference](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Cloud-Backups/operation/getDataProtectionSettings) and [Backup Compliance Policy Prohibited Actions](https://www.mongodb.com/docs/atlas/backup/cloud-backup/backup-compliance-policy/#prohibited-actions)

Reviewer 3: Minor nitpick, feel free to take or leave
```suggestion
For more information, see [MongoDB Atlas API Reference](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Cloud-Backups/operation/getDataProtectionSettings) and [Backup Compliance Policy Prohibited Actions](https://www.mongodb.com/docs/atlas/backup/cloud-backup/backup-compliance-policy/#prohibited-actions)
```

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: backup_compliance_policy"
sidebar_current: "docs-mongodbatlas-datasource-backup-compliance-policy"
description: |-
    Provides a Backup Compliance Policy Datasource.
---

# Data Source: mongodbatlas_backup_compliance_policy

> `mongodbatlas_backup_compliance_policy` provides a Backup Compliance Policy datasource. An Atlas Backup Compliance Policy provides the current protection policy settings for the project. 

Reviewer 2: ```suggestion
`mongodbatlas_backup_compliance_policy` provides a Backup Compliance Policy data source. An Atlas Backup Compliance Policy provides the current protection policy settings for the project. 
```

}

resource "mongodbatlas_backup_compliance_policy" "backup_policy" {
  project_id                 = "<PROJECT-ID>"
  authorized_email           = "user@email.com"
  copy_protection_enabled    = false
  pit_enabled                = false
  encryption_at_rest_enabled = false

  restore_window_days = 7

  on_demand_policy_item {

    frequency_interval = 0
    frequency_type     = "ondemand"
    retention_unit     = "days"
    retention_value    = 3
  }
  
  scheduled_policy_items {
    
      frequency_interval = 6
      frequency_type     = "hourly"
      retention_unit     = "days"
      retention_value    = 7
    }

scheduled_policy_items {
    
      frequency_interval = 0
      frequency_type     = "daily"
      retention_unit     = "days"
      retention_value    = 7
    }

scheduled_policy_items {
    
      frequency_interval = 0
      frequency_type     = "weekly"
      retention_unit     = "weeks"
      retention_value    = 4
    }

scheduled_policy_items {
    
      frequency_interval = 0
      frequency_type     = "monthly"
      retention_unit     = "months"
      retention_value    = 12
    }

}
```

## Argument Reference

> * `project_id` - (Required) The unique identifier of the project for the Atlas cluster.

Reviewer 2: ```suggestion
* `project_id` - (Required) Unique 24-hexadecimal digit string that identifies your project
```

> retrieve.

Reviewer 2: ```suggestion

```

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `authorized_email` - (Email address of the user who authorized to updated the Backup Compliance Policy settings.

Reviewer 2: ```suggestion
* `authorized_email` - Email address of the user who authorized to updated the Backup Compliance Policy settings.
```

* `on_demand_policy_item.#.policy_item.#.frequency_interval` - The frequency interval for a set of snapshots.
* `on_demand_policy_item.#.policy_item.#.frequency_type` - A type of frequency (hourly, daily, weekly, monthly).
* `on_demand_policy_item.#.policy_item.#.retention_unit` - The unit of time in which snapshot retention is measured (days, weeks, months).
* `policies.#.policy_item.#.retention_value` - The number of days, weeks, or months the snapshot is retained.
* 
#### Scheduled Policy Items
* `scheduled_policy_items.#.policy_item` - A list of specifications for a policy.
* `scheduled_policy_items.#.policy_item.#.id` - Unique identifier for this policy item.
* `scheduled_policy_items.#.policy_item.#.frequency_interval` - The frequency interval for a set of snapshots.
* `scheduled_policy_items.#.policy_item.#.frequency_type` - A type of frequency (hourly, daily, weekly, monthly).
* `scheduled_policy_items.#.policy_item.#.retention_unit` - The unit of time in which snapshot retention is measured (days, weeks, months).
* `scheduled_policy_items.#.policy_item.#.retention_value` - The number of days, weeks, or months the snapshot is retained.

> For more information see: [MongoDB Atlas API Reference.](hhttps://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Cloud-Backups/operation/getDataProtectionSettings)

Reviewer 2: ```suggestion
For more information see: [MongoDB Atlas API Reference](https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Cloud-Backups/operation/getDataProtectionSettings) and [Backup Compliance Policy Prohibited Actions](https://www.mongodb.com/docs/atlas/backup/cloud-backup/backup-compliance-policy/#prohibited-actions)
```

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: backup_compliance_policy"
sidebar_current: "docs-mongodbatlas-datasource-backup-compliance-policy"
description: |-
    Provides a Backup Compliance Policy Datasource.
---

# Data Source: mongodbatlas_backup_compliance_policy

> `mongodbatlas_backup_compliance_policy` provides a Backup Compliance Policy data source. An Atlas Backup Compliance Policy provides the current protection policy settings for the project. Prevent any user, regardless of role, from modifying or deleting specific cluster configurations and backups. When enabled, the Backup Compliance Policy will be applied as the minimum policy for all clusters and can only be disabled by contacting MongoDB support. Only supported for clusters M10 or higher.

Reviewer 3: Suggestions to simplify and hopefully clarify
```suggestion
`mongodbatlas_backup_compliance_policy` provides an Atlas Backup Compliance Policy. An Atlas Backup Compliance Policy contains the current protection policy settings for a project. A compliance policy prevents any user, regardless of role, from modifying or deleting specific cluster configurations and backups. To disable a Backup Compliance Policy, you must contact MongoDB support. Backup Compliance Policies are only supported for clusters M10 and higher and are applied as the minimum policy for all clusters.
```

> -> **NOTE:** Groups and projects are synonymous terms. You may find `groupId` in the official documentation.

Reviewer 2: ```suggestion
-> **IMPORTANT NOTE:** Once you enable a Backup Compliance Policy, no user, regardless of role, can disable the Backup Compliance Policy via Terraform, or any other method, without contacting MongoDB support. This means that, once enabled, some resources defined in Terraform can not be modified. To learn more, see the full list of [Backup Compliance Policy Prohibited Actions and Considerations](https://www.mongodb.com/docs/atlas/backup/cloud-backup/backup-compliance-policy/#configure-a-backup-compliance-policy).

-> **NOTE:** Groups and projects are synonymous terms. You might find `groupId` in the official documentation.
```

Reviewer 3: To add a few copy edits on top of @Zuhairahmed 's suggestion:

Once you enable a Backup Compliance Policy, no user, regardless of role, can disable the Backup Compliance Policy via Terraform, or any other method, without contacting MongoDB support. This means that, once enabled, some resources defined in Terraform can not be modified. To learn more, see the full list of [Backup Compliance Policy Prohibited Actions and Considerations](https://www.mongodb.com/docs/atlas/backup/cloud-backup/backup-compliance-policy/#configure-a-backup-compliance-policy).

and also:
("may" changed to "might" because the MongoDB style guide prefers "may" for permission and "might" for chance)

Groups and projects are synonymous terms. You might find `groupId` in the official documentation.

Reviewer 2: this is awesome one! thanks high level of detail here @zach-carr!

## Example Usage

```terraform
resource "mongodbatlas_cluster" "my_cluster" {
  project_id   = "<PROJECT-ID>"
  name         = "clusterTest"

  //Provider Settings "block"
  provider_name               = "AWS"
  provider_region_name        = "EU_CENTRAL_1"
  provider_instance_size_name = "M10"
  provider_backup_enabled     = true // enable cloud backup snapshots
}

> resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {

Reviewer 2: @martinstibbe this is deprecated resource, can you update like how we did in resource example?

  project_id   = mongodbatlas_cluster.my_cluster.project_id
  cluster_name = mongodbatlas_cluster.my_cluster.name

  reference_hour_of_day    = 3
  reference_minute_of_hour = 45
  restore_window_days      = 4

>   policies {

Reviewer 2: @martinstibbe these policy parameters have now changed, can you update like how we did in resource docs example?

  project_id   = mongodbatlas_cloud_backup_schedule.test.project_id
  cluster_name = mongodbatlas_cloud_backup_schedule.test.cluster_name
}

data "mongodbatlas_backup_compliance_policy" "backup_policy" {
  project_id = mongodbatlas_cloud_backup_schedule.test.id
}

resource "mongodbatlas_backup_compliance_policy" "backup_policy" {
  project_id                 = "<PROJECT-ID>"
  authorized_email           = "user@email.com"
  copy_protection_enabled    = false
  pit_enabled                = false
  encryption_at_rest_enabled = false

  restore_window_days = 7

  on_demand_policy_item {
		  frequency_interval = 0
		  retention_unit     = "days"
		  retention_value    = 3
		}
		
		policy_item_hourly {
			frequency_interval = 6
			retention_unit     = "days"
			retention_value    = 7
		  }
	  
		policy_item_daily {
			frequency_interval = 0
			retention_unit     = "days"
			retention_value    = 7
		  }
	  
		  policy_item_weekly {
			frequency_interval = 0
			retention_unit     = "weeks"
			retention_value    = 4
		  }
	  
		  policy_item_monthly {
			frequency_interval = 0
			retention_unit     = "months"
			retention_value    = 12
		  }

}
```

## Argument Reference

> * `project_id` - (Required) Unique 24-hexadecimal digit string that identifies your project

Reviewer 2: ```suggestion
* `project_id` - (Required) Unique 24-hexadecimal digit string that identifies your project.
* `authorized_email` - Email address of a security or legal representative for the Backup Compliance Policy who is authorized to update the Backup Compliance Policy settings.
* `copy_protection_enabled` - Flag that indicates whether to enable additional backup copies for the cluster. If unspecified, this value defaults to false.
* `pit_enabled` - Flag that indicates whether the cluster uses Continuous Cloud Backups with a Backup Compliance Policy. If unspecified, this value defaults to false.
* `encryption_at_rest_enabled` - Flag that indicates whether Encryption at Rest using Customer Key Management is required for all clusters with a Backup Compliance Policy. If unspecified, this value defaults to false.
* `restore_window_days` - Number of previous days that you can restore back to with Continuous Cloud Backup with a Backup Compliance Policy. You must specify a positive, non-zero integer, and the maximum retention window can't exceed the hourly retention time. This parameter applies only to Continuous Cloud Backups with a Backup Compliance Policy.
```

Reviewer 3: Suggestions on top of @Zuhairahmed 's:

Just added a period to the end:
`* `project_id` - (Required) Unique 24-hexadecimal digit string that identifies your project.
`

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `authorized_email` - Email address of the user who authorized to updated the Backup Compliance Policy settings.

Reviewer 3: ```suggestion
* `authorized_email` - Email address of the user who is authorized to update the Backup Compliance Policy settings.
```

* `copy_protection_enabled` - Flag that indicates whether to enable additional backup copies for the cluster. If unspecified, this value defaults to false.
* `pit_enabled` - Flag that indicates whether the cluster uses Continuous Cloud Backups with a Backup Compliance Policy. If unspecified, this value defaults to false.
* `encryption_at_rest_enabled` - Flag that indicates whether Encryption at Rest using Customer Key Management is required for all clusters with a Backup Compliance Policy. If unspecified, this value defaults to false.

> * `reference_minute_of_hour` - UTC Minute of day between 0 and 59 representing which minute of the referenceHourOfDay that Atlas takes the snapshot.

Reviewer 3: Edit: this is just a suggestion, please ignore if you want to be specific about UTC. In that case, just lowercase minute ("UTC minute of day") 

```suggestion
* `reference_minute_of_hour` - Integer between 0 and 59 representing which minute of the referenceHourOfDay that Atlas takes the snapshot.
```
